src/common/Validation.ml

```diff
@@ -63,7 +58,6 @@ end
   let apply x ~f = 
     match (f,x) with
   ```
-
   `apply` 'unwraps' the function and the value and performs different actions 
   depending on whether an error has occured. Because we have two arguments each 
   of which can be in two states, we have to check for four conditions:
```

Comment:
It says here (and in the railway oriented programming blog post) that we generally want to exit early if we hit a failure. How does this pattern change to account for collecting all errors and warnings?

Reply:
> that we generally want to exit early if we hit a failure

That's the case with the _monadic_ interface to `Result.t` - as soon as we hit an error case we know that subsequent computations that depend on it cannot be `Ok`. You can see this from the type of `bind`:

```ocaml
val bind : `a result -> f:('a -> 'b result) -> 'b result
``` 
The second function argument (representing the continuation of a computation) takes an `'a` and the only way it can get is the first argument is `Ok` and contains an '`a'. If it's an `Error` the computation cannot proceed*.

However, ultimately we want `Validation` to expose only an `Applicative` interface (see previous reply). Importantly the implementation of `apply` for `Validation` is _not_ the same as `Result`. For a `Result.t` type we don't know what to do if we have two `Error`s but in `Validation.t` we know we can combine the data in the `Error` case. So, for `Validation`, given an `Applicative` interface, we still want to proceed with the other 'parallel' computations.

(* I have seen this short-circuiting implemented with exceptions in OCaml but more usually you just define `bind` such that it performs the redundant computation.)

Reply:
Gotcha. The fact that we both 1) need sequencing but 2) would like to not stop computation the second something in the sequence fails makes me wonder if we actually don't want either one of Monad or Applicative here. It seems like we basically want a Writer monad around a `(Warning.t * Error.t) list`, maybe?